%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Please note that whilst this template provides a
% preview of the typeset manuscript for submission, it
% will not necessarily be the final publication layout.
%
% letterpaper/a4paper: US/UK paper size toggle
% num-refs/alpha-refs: numeric/author-year citation and bibliography toggle

%\documentclass[letterpaper]{oup-contemporary}
\documentclass[a4paper,num-refs]{oup-contemporary}

%%% Journal toggle; only specific options recognised.
%%% (Only "gigascience" and "general" are implemented now. Support for other journals is planned.)
\journal{gigascience}

\usepackage{graphicx}
\usepackage{siunitx}
\usepackage{listings}

%%% Flushend: You can add this package to automatically balance the final page, but if things go awry (e.g. section contents appearing out-of-order or entire blocks or paragraphs are coloured), remove it!
% \usepackage{flushend}

\title{Training Infrastructure as a Service}

%%% Use the \authfn to add symbols for additional footnotes, if any. 1 is reserved for correspondence emails; then continuing with 2 etc for contributions.
\author[1,\authfn{1},\authfn{2}]{Helena~Rasche}
\author[1,\authfn{2}]{Bj\"orn~Gr\"uning}

\affil[1]{Bioinformatics Group, Department of Computer Science, University of Freiburg, 79110 Freiburg im Breisgau, Germany}

%%% Author Notes
\authnote{\authfn{1}helena.rasche@gmail.com}
\authnote{\authfn{2}Contributed equally.}

%%% Paper category
\papercat{Technical Note}

%%% "Short" author for running page header
\runningauthor{Rasche and Gr\"uning}

%%% Should only be set by an editor
\jvolume{00}
\jnumber{0}
\jyear{2020}

\begin{document}

\begin{frontmatter}
\maketitle
\begin{abstract}
%The Abstract (250 words maximum) should be structured to include the following details: \textbf{Background}, the context and purpose of the study; \textbf{Results}, the main findings; \textbf{Conclusions}, brief summary and potential implications. Please minimize the use of abbreviations and do not cite references in the abstract.
\textbf{Background:} Hands-on training, whether it is in Bioinformatics or other scientific domains, requires significant resources and knowledge to setup and run.
Trainers must have access to infrastructure that can support the sudden spike in usage, with classes of 30 or more trainees simultaneously running resource intensive tools.
For efficient classes, the jobs must run quickly, without queuing delays, lest they disrupt the timetable set out for the class. Often times this is achieved via running on a private server where there is no contention for the queue, and therefore no or minimal waiting time. However, this requires the teacher or trainer to have the technical knowledge to manage compute infrastructure, in addition to their didactic responsibilities. This presents significant burdens to potential training events, in terms of infrastructure cost, person-hours of preparation, technical knowledge, and available staff to manage such events.

\textbf{Findings:} Galaxy Europe has developed Training Infrastructure as a Service (TIaaS) which we provide to the scientific commnuity as a service built on top of the Galaxy Platform. Training event organisers request a training and Galaxy administrators can allocate private queues specifically for the training. Trainees are transparently placed in a private queue where their jobs run without delay. Trainers access the dashboard of the TIaaS Service and can remotely follow the progress of their trainees without in-person interactions.

\textbf{Conclusions:} TIaaS on Galaxy Europe provides reusable and fast infrastructure for Galaxy training. The instructor dashboard provides visibility into class progress, making in-person trainings more efficient and remote training possible. In the past 24 months, $>110$ trainings with over 3000 trainees have used this infrastructure for training, across scientific domains, all enjoying the accessibility and reproducibility of Galaxy for training the next generation of bioinformaticians. TIaaS itself is an extension to Galaxy which can be deployed by any Galaxy administrator to provide similar benefits for their users. \url{https://galaxyproject.eu/tiaas}
\end{abstract}

\begin{keywords}
Galaxy; Training; Teaching
\end{keywords}
\end{frontmatter}

\section{Findings}
\subsection{Background}

% The background section should be written in a way that is accessible to researchers without specialist knowledge in that area and must clearly state---and, if helpful, illustrate---the background to the research and its aims. The section should end with a brief statement of what is being reported in the article.

% todo: sentence about how importnt hands-on training is?
With the volume of bioinformatics data becoming available, the availability of training for bioinformaticians is not keeping up (\cite{Attwood2017}).
The Galaxy platform \cite{afgan2018galaxy} provides one such infrastructure on which to conduct trainings, it provides a user-friendly web-based interface to command line tools. With a wide range of tools across bioinformatics domains and beyond, and pre-existing popularity within the life sciences community, it is an ideal platform for training (\cite{gtn}).

The Galaxy community has developed a wide array of hands-on training materials covering the bioinformatics domains and beyond (\cite{training-site}) in an attempt to combat this issue, but in order to run these tutorials at scale, one occasionally needs access to significant resources. The quite popular ``Reference-based RNA-Seq data analysis'' tutorial uses the STAR aligner (\cite{Dobin2012}), and while an ultra-fast aligner is ideal during training, it also consumes 32-64 GB of RAM at minimum. Running a single STAR job might be fine, but if 30 trainees run STAR simultaneously, it can present unexpected delays due to jobs contending for available compute resources. Most Galaxy servers submit jobs to compute clusters, but these clusters must also meet the needs of all of theother users of the Galaxy servers, and might not have enough resources for the requested jobs. Jobs thus queue and wait until resources become available.

Many tools across Galaxy have similarly large resource requirements, requirements which are only discoverable through testing. If trainers choose to manage their own infrastructure they need to obtain this knowledge via testing, and then configure their infrastructure to support the minimum memory and CPU requirements. Alternatively if the instructor chooses to use one of the many public Galaxy servers, they can rely on those services, but their trainees' jobs could unexpectedly sit in shared queues for long periods of time. This is a significant barriers for potential, motivated trainers who wish to give a Galaxy or bioinformatics training courses.

We have developed Training Infrastructure as a Service (TIaaS) initially to make our own courses run more smoothly due to previous experiences with the aforementioned problems. TIaaS removes the main bottlenecks of training: setting up the training infrastructure is nearly completely automated, jobs are guaranteed to have private resources available so they can always run without delay, and it ensures separation of responsibilities between trainers who are teaching and the server administrators responsible for Galaxy. We have shared this service with the Galaxy training community to very positive feedback.

\subsection{Results}
\begin{figure}[bt!]
\centering
\includegraphics[width=\linewidth]{images/dashboard.png}
\includegraphics[width=\linewidth]{images/queue.png}
\caption{The top of the training dashboard page shows the status of the jobs in the past hours. A heatmap of the tools which were run indiciates if everything is running smoothly or if there is anything the trainer should look into. As trainees follow along and run different tools these show up immediately, allowing trainers to identify if everyone has started or finished a specific step.. The bottom image shows the rest of the training dashboard, which lists jobs that were run chronologically, colour coded first by user, and second by the job status. Random identifiers are used to protect user privacy as we wanted dashboards to be public.}\label{figure:dashboard}
\end{figure}

We implemented a Django based web service and job scheduling rules which function together to present a private queue for users in specific Galaxy user groups. This system can be generically useful for temporary and private compute resources, but is currently focused on training use case. Trainers register their course and teaching materials with an online form. Galaxy administrators review requests, using information about their class size, the tools used in those training materials, and the resource allocations of those tools on the infrastructure to estimate the required compute resources.

If space is available and any other criteria is met, the training can then be approved. Next, administrators optionally deploy private compute resources in a local cloud and attach them to their existing Galaxy scheduling. Administrators are provided with a URL such as \url{https://usegalaxy.eu/join-training/test}, which they share with the trainer, and the trainer can then in turn provide to their trainees. When training participants access the URL, they are registered in the TIaaS system, without the service administrator or trainers needing to be aware of users identities for optimal GDPR compliance. The scheduler, now aware of the training group, will grant any job run by someone in that group the ability to run on the private training nodes (Figure \ref{figure:queue}).

\begin{figure}[bt!]
\centering
\includegraphics[width=\linewidth]{images/rules.png}
\caption{All jobs are processed by the same Galaxy server, but jobs from users in the training groups receive special handling. These jobs are allowed to run on the protected training resources (purple). If the training resource is full, these jobs can spill over to the main queue if necessary.}\label{figure:queue}
\end{figure}

The course dashboard, visualising the progress of participants (Figure \ref{figure:dashboard}), has significantly eased the lives of trainers. The dashboard provides instantaneous, aggregated, and anonymized feedback for the trainers into how their trainees are progressing in a GDPR compliant manner. It has also enabled remote and hybrid trainings to take place, where trainers present a topic via video and it is broadcast to numerous individuals or sites. Requiring students or teaching sites to constantly provide up-to-date information on how participants are progressing simply does not work. With the training dashboard this is unnecessary as trainers, helpers, and even trainees can see how everyone is progressing online, instantaneously. Teachers can see if steps are completed successfully on average, or if there are many issues and they might need to pause or explain the step in more detail.

\subsubsection{Usage}
Since the introduction of TIaaS in July 2018, it has seen nearly constant use with more than 110 trainings occuring on the platform, all across the world (Figure \ref{figure:map}). Everything from one day workshops for bioinformaticians to multi-month courses for highschool students have been hosted by TIaaS and UseGalaxy.eu, covering topics such as HTS, RNA-Seq, ChIP-Seq, Imaging Analysis, Proteomics, and Machine Learning.

\begin{figure}[bt!]
\centering
\includegraphics[width=\linewidth]{images/map.png}
	\caption{Since its introduction, TIaaS has seen worldwide usage despite--or perhaps in some cases due to--the server being hosted in Germany. This dashboard is part of the TIaaS platform, allowing deployers to share their numbers with funding agencies and the world.}\label{figure:map}
\end{figure}

Class sizes have ranged considerably, from the median of 15 participants to a maximum of 80 at a Transcriptomics workshop. Most course were quite short training events with a median of two days, however some ran for multiple months like a highschool course that used the service for three months, over the course of a semester, teaching Metagenomics. As the system is quite generic, it can easily accomodate a range of use cases.

\section{Methods}

\subsection{Implementation}
TIaaS was written in Python with the Django framework. It has been designed from the start to have a very limited scope: provide a form to register events, an approval flow for administrators, and a connection to the Galaxy database to manage user-group relationships when needed.

For instructors a form is provided (\texttt{/tiaas/new}) which permits them to register a new training event. When submitted, this goes into the associated database. Administrators can view the requested training events and approve or reject them using the built in Django admin interface. When users visit their their training url (\texttt{/join-training/<id>}) the system accesses their Galaxy session cookie, and decodes it. They are automatically registered as part of a Galaxy group named after the training (e.g. \texttt{training-<id>}) which is created as needed.

When instructors or students visit the dashboard (Figure \ref{figure:dashboard}), the training ID is extracted from the URL (e.g. ``test'' from \url{https://usegalaxy.eu/join-training/test/status}), and all non-terminal jobs, in the past 1-6 hours, from those users are presented in a de-identified manner.

When a job is submitted by a user in a training group, UseGalaxy.eu's job scheduling system system reads the user's groups and roles, and if any of these include something prefixed with \texttt{training-}, then this is converted to a job scheduler specific requirement string (Figure \ref{code:scheduler}). Ideally these are constructured to prefer training nodes, and spill over to the main queue if training nodes are full, but this feature is dependant on specific scheduler capabilities.

\begin{figure}[!ht]
\centering
\begin{lstlisting}[frame=single,language=Python]  % Start your code-block
Thisisignored?
def process_job(job, user):
    resources = ['main-cluster']
    rank = 'main-cluster'
    if inTrainingGroup(user):
        training = getTrainingGroup(user)
        resources.append(training)
        rank = training

    job.resources = resources
    job.rank = rank
    return job
\end{lstlisting}
\caption{Python pseudocode representing how to process TIaaS jobs and allocate them to a private queue.\label{code:scheduler}}
\end{figure}

\subsection{Deploying}
As the Galaxy community has largely settled on Ansible for deployment of Galaxy, and related components, an Ansible role was produced for deploying the TIaaS Service.

\section{Availability of source code and requirements}

\begin{itemize}
\item Project name: ~Training Infrastructure as a Service
\item Github repository:~\url{https://github.com/usegalaxy-eu/tiaas2/}
\item Training Manual: ~\url{https://training.galaxyproject.org/training-material/topics/instructors/tutorials/setup-tiaas-for-training/tutorial.html}
\item Operating system(s): ~Unix
\item Other requirements: ~Galaxy version 18.01 or higher
\item License: ~GNU AGPL-3.0
\end{itemize}

\section{Availability of supporting data and materials}
All code is open source and available on GitHub (https://github.com/usegalaxy-eu/tiaas2).

\section{Declarations}

\subsection{List of abbreviations}
\begin{itemize}
\item TIaaS: Training Infrastructure as a Service
\end{itemize}


\subsection{Competing Interests}
The authors declare that they have no competing interests.

\subsection{Funding}
This project was made possible with the support of the Albert Ludwig University of Freiburg.
The work is in part funded by Collaborative Research Centre 992 Medical Epigenetics (DFG grant SFB 992/1 2012) and German Federal Ministry of Education and Research (BMBF grants 031 A538A/A538C RBC and 031L0101B/031L0101C de.NBI-epi). The article processing charge was funded by the Baden-Württemberg Ministry of Science, Research and Art and the University of Freiburg in the funding programme Open Access Publishing.

\subsection{Author's Contributions}
HR and BG conceived of the presented idea. HR carried out the implementation. HR and BG contributed to the writing of the manuscript.

\section{Acknowledgements}
The authors would like to thank the Galaxy community for their enthusiasm for this project, and their feedback on each iteration.

%% Specify your .bib file name here, without the extension
\bibliography{paper-refs}

\end{document}
